sdk/lib/cmlib/cmcheck.c

```diff
@@ -0,0 +1,341 @@
+/*
+ * PROJECT:         ReactOS Kernel
+ * LICENSE:         GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:         Configuration Manager Library - Registry Validation
+ * COPYRIGHT:       Copyright 2022 George Bișoc <george.bisoc@reactos.org>
+ */
+
+/* INCLUDES ******************************************************************/
+
+#include "cmlib.h"
+#define NDEBUG
+#include <debug.h>
+
+/* DEFINES  ******************************************************************/
+
+#define GET_HHIVE(CmHive) (&((CmHive)->Hive))
+#define GET_HHIVE_ROOT_CELL(Hive) ((Hive)->BaseBlock->RootCell)
+#define GET_HHIVE_BIN(Hive, StorageIndex, BlockIndex) ((PHBIN)Hive->Storage[StorageIndex].BlockList[BlockIndex].BinAddress)
+#define GET_CELL_BIN(Bin) ((PHCELL)(Bin + 1))
+
+#if !defined(CMLIB_HOST) && !defined(_BLDR_)
+extern PCMHIVE CmiVolatileHive;
+#endif
+
+/* PRIVATE FUNCTIONS **********************************************************/
+
+CM_CHECK_REGISTRY_STATUS
+CmpValidateKey(
+    _In_ PHHIVE Hive,
+    _In_ BOOLEAN SecurityDefaulted,
+    _In_ HCELL_INDEX ParentCell,
+    _In_ HCELL_INDEX Cell,
+    _In_ ULONG Flags)
+{
+    // TODO
+    return 0;
+}
+
+CM_CHECK_REGISTRY_STATUS
+CmpValidateRegistryInternal(
+    _In_ PHHIVE Hive,
+    _In_ ULONG Flags,
+    _In_ BOOLEAN SecurityDefaulted)
+{
+    CM_CHECK_REGISTRY_STATUS CmStatusCode;
+    HCELL_INDEX SubKeyCell;
+    PCM_KEY_NODE RootNode;
+    ULONG SubKeyIndex;
+    HCELL_INDEX Cell, RootCell, ParentCell, Sibling;
+
+    /*
+     * Start from the root cell and check that
+     * the root itself is OK in the first place.
+     */
+    ParentCell = HCELL_NIL;
+    RootCell = GET_HHIVE_ROOT_CELL(Hive);
+    ASSERT(RootCell);
+
+    CmStatusCode = CmpValidateKey(Hive,
+                                  SecurityDefaulted,
+                                  ParentCell,
+                                  RootCell,
+                                  Flags);
+    if (!CM_CHECK_REGISTRY_SUCCESS(CmStatusCode))
+    {
+        /* TODO: We must try to repair the hive here if possible */
+        DPRINT1("CmpValidateRegistryInternal(): The root cell is corrupt (cell %lu)!", RootCell);
+        return CmStatusCode;
+    }
+
+    /* Obtain the root node */
+    RootNode = (PCM_KEY_NODE)HvGetCell(Hive, RootCell);
+    if (!RootNode)
+    {
+        DPRINT1("CmpValidateRegistryInternal(): Couldn't obtain the root node of root cell (cell %lu)!\n", RootCell);
+        return CM_CHECK_REGISTRY_NO_ROOT_NODE;
+    }
+
+    /*
+     * Begin looping and validate each subkey now.
+     * From there we also check the order consistency
+     * (aka lexicographical order).
+     */
+    ParentCell = RootCell;
+    Sibling = HCELL_NIL;
+    for (SubKeyIndex = 0; SubKeyIndex < RootNode->SubKeyCounts[Stable]; SubKeyIndex++)
+    {
+        /* Obtain the subkey at the index we're in */
+        SubKeyCell = CmpFindSubKeyByNumber(Hive, RootNode, SubKeyIndex);
```

Comment:
Not sure you can use this function. It expects that the referenced data is valid (and would assert/crash if not); however here, we are validating and repairing possibly corrupted data.